tsdb/head.go

```diff
@@ -2013,13 +2056,66 @@ func (s *memSeries) cleanupAppendIDsBelow(bound uint64) {
 	}
 }
 
-func (s *memSeries) head() *memChunk {
-	return s.headChunk
-}
-
 type memChunk struct {
 	chunk            chunkenc.Chunk
 	minTime, maxTime int64
+	next             *memChunk // link to the next element on the list
+}
+
+// len returns the length of memChunk list, including the element it was called on.
+func (mc *memChunk) len() (count int) {
+	elem := mc
+	for elem != nil {
+		count++
+		elem = elem.next
+	}
+	return count
+}
+
+// last returns the last element on the list.
+// For single element list this will be the same memChunk last() was called on
+func (mc *memChunk) last() (elem *memChunk) {
+	elem = mc
+	for elem.next != nil {
+		elem = elem.next
+	}
+	return elem
+}
+
+// atOffset returns a memChunk that's Nth element on the linked list.
+func (mc *memChunk) atOffset(offset int) (elem *memChunk) {
+	if offset == 0 {
+		return mc
+	}
+	if offset < 0 {
+		return nil
+	}
+
+	var i int
+	elem = mc
+	for i < offset {
+		i++
+		elem = elem.next
+		if elem == nil {
+			break
+		}
+	}
+
+	return elem
+}
+
+// toReversedSlice returns a slice of memChunks where first element of the slice
+// is the last element on the linked list.
+func (mc *memChunk) toReversedSlice() []*memChunk {
```

Comment:
Since having only 1 chunk is the expected outcome for majority cases, we can shortcut that condition by checking `mc.next` for `nil` and avoid some functions calls.